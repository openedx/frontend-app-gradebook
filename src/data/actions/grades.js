import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import {
  STARTED_FETCHING_GRADES,
  FINISHED_FETCHING_GRADES,
  ERROR_FETCHING_GRADES,
  GOT_GRADES,
  GRADE_UPDATE_REQUEST,
  GRADE_UPDATE_SUCCESS,
  GRADE_UPDATE_FAILURE,
  TOGGLE_GRADE_FORMAT,
  FILTER_BY_ASSIGNMENT_TYPE,
  OPEN_BANNER,
  CLOSE_BANNER,
  START_UPLOAD,
  UPLOAD_COMPLETE,
  UPLOAD_ERR,
  GOT_BULK_HISTORY,
  BULK_HISTORY_ERR,
  GOT_GRADE_OVERRIDE_HISTORY,
  DONE_VIEWING_ASSIGNMENT,
  ERROR_FETCHING_GRADE_OVERRIDE_HISTORY,
  UPLOAD_OVERRIDE,
  UPLOAD_OVERRIDE_ERROR,
  BULK_GRADE_REPORT_DOWNLOADED,
  INTERVENTION_REPORT_DOWNLOADED,
} from '../constants/actionTypes/grades';
import GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG from '../constants/errors';
import LmsApiService from '../services/LmsApiService';
import { sortAlphaAsc, formatDateForDisplay } from './utils';
import {
  formatMaxAssignmentGrade, formatMinAssignmentGrade, formatMaxCourseGrade, formatMinCourseGrade,
} from '../selectors/grades';
import { getFilters } from '../selectors/filters';

const defaultAssignmentFilter = 'All';

const startedCsvUpload = () => ({ type: START_UPLOAD });
const finishedCsvUpload = () => ({ type: UPLOAD_COMPLETE });
const csvUploadError = data => ({ type: UPLOAD_ERR, data });
const gotBulkHistory = data => ({ type: GOT_BULK_HISTORY, data });
const bulkHistoryError = () => ({ type: BULK_HISTORY_ERR });

const startedFetchingGrades = () => ({ type: STARTED_FETCHING_GRADES });
const finishedFetchingGrades = () => ({ type: FINISHED_FETCHING_GRADES });
const errorFetchingGrades = () => ({ type: ERROR_FETCHING_GRADES });
const errorFetchingGradeOverrideHistory = errorMessage => ({
  type: ERROR_FETCHING_GRADE_OVERRIDE_HISTORY,
  errorMessage,
});

const gotGrades = ({
  grades, cohort, track, assignmentType, headings, prev,
  next, courseId, totalUsersCount, filteredUsersCount,
}) => ({
  type: GOT_GRADES,
  grades,
  cohort,
  track,
  assignmentType,
  headings,
  prev,
  next,
  courseId,
  totalUsersCount,
  filteredUsersCount,
});

const gotGradeOverrideHistory = ({
  overrideHistory, currentEarnedAllOverride, currentPossibleAllOverride,
  currentEarnedGradedOverride, currentPossibleGradedOverride,
  originalGradeEarnedAll, originalGradePossibleAll, originalGradeEarnedGraded,
  originalGradePossibleGraded,
}) => ({
  type: GOT_GRADE_OVERRIDE_HISTORY,
  overrideHistory,
  currentEarnedAllOverride,
  currentPossibleAllOverride,
  currentEarnedGradedOverride,
  currentPossibleGradedOverride,
  originalGradeEarnedAll,
  originalGradePossibleAll,
  originalGradeEarnedGraded,
  originalGradePossibleGraded,
});

const gradeUpdateRequest = () => ({ type: GRADE_UPDATE_REQUEST });
const gradeUpdateSuccess = (courseId, responseData) => ({
  type: GRADE_UPDATE_SUCCESS,
  courseId,
  payload: { responseData },
});
const gradeUpdateFailure = (courseId, error) => ({
  type: GRADE_UPDATE_FAILURE,
  courseId,
  payload: { error },
});
const uploadOverrideSuccess = courseId => ({
  type: UPLOAD_OVERRIDE,
  courseId,
});
// This action for google analytics only. Doesn't change redux state.
const downloadBulkGradesReport = courseId => ({
  type: BULK_GRADE_REPORT_DOWNLOADED,
  courseId,
});
// This action for google analytics only. Doesn't change redux state.
const downloadInterventionReport = courseId => ({
  type: INTERVENTION_REPORT_DOWNLOADED,
  courseId,
});
const uploadOverrideFailure = (courseId, error) => ({
  type: UPLOAD_OVERRIDE_ERROR,
  courseId,
  payload: { error },
});

const toggleGradeFormat = formatType => ({ type: TOGGLE_GRADE_FORMAT, formatType });

const filterAssignmentType = filterType => (
  dispatch => dispatch({
    type: FILTER_BY_ASSIGNMENT_TYPE,
    filterType,
  })
);

const openBanner = () => ({ type: OPEN_BANNER });
const closeBanner = () => ({ type: CLOSE_BANNER });

const fetchGrades = (
  courseId,
  cohort,
  track,
  assignmentType,
  options = {},
) => (
  (dispatch, getState) => {
    dispatch(startedFetchingGrades());
    const {
      assignment,
      assignmentGradeMax: assignmentMax,
      assignmentGradeMin: assignmentMin,
      courseGradeMin,
      courseGradeMax,
    } = getFilters(getState());
    const { id: assignmentId } = assignment || {};
    const assignmentGradeMax = formatMaxAssignmentGrade(assignmentMax, { assignmentId });
    const assignmentGradeMin = formatMinAssignmentGrade(assignmentMin, { assignmentId });
    const courseGradeMinFormatted = formatMinCourseGrade(courseGradeMin);
    const courseGradeMaxFormatted = formatMaxCourseGrade(courseGradeMax);
    return LmsApiService.fetchGradebookData(
      courseId,
      options.searchText || null,
      cohort,
      track,
      {
        assignment: assignmentId,
        assignmentGradeMax,
        assignmentGradeMin,
        courseGradeMin: courseGradeMinFormatted,
        courseGradeMax: courseGradeMaxFormatted,
      },

    )
      .then(response => response.data)
      .then((data) => {
        dispatch(gotGrades({
          grades: data.results.sort(sortAlphaAsc),
          cohort,
          track,
          assignmentType,
          prev: data.previous,
          next: data.next,
          courseId,
          totalUsersCount: data.total_users_count,
          filteredUsersCount: data.filtered_users_count,
        }));
        dispatch(finishedFetchingGrades());
        if (options.showSuccess) {
          dispatch(openBanner());
        }
      })
      .catch(() => {
        dispatch(errorFetchingGrades());
      });
  }
);

const formatGradeOverrideForDisplay = historyArray => historyArray.map(item => ({
  date: formatDateForDisplay(new Date(item.history_date)),
  grader: item.history_user,
  reason: item.override_reason,
  adjustedGrade: item.earned_graded_override,
}));

const doneViewingAssignment = () => dispatch => dispatch({
  type: DONE_VIEWING_ASSIGNMENT,
});
const fetchGradeOverrideHistory = (subsectionId, userId) => (
  dispatch => LmsApiService.fetchGradeOverrideHistory(subsectionId, userId)
    .then(response => response.data)
    .then((data) => {
      if (data.success) {
        dispatch(gotGradeOverrideHistory({
          overrideHistory: formatGradeOverrideForDisplay(data.history),
          currentEarnedAllOverride: data.override ? data.override.earned_all_override : null,
          currentPossibleAllOverride: data.override ? data.override.possible_all_override : null,
          currentEarnedGradedOverride: data.override ? data.override.earned_graded_override : null,
          currentPossibleGradedOverride: data.override
            ? data.override.possible_graded_override : null,
          originalGradeEarnedAll: data.original_grade ? data.original_grade.earned_all : null,
          originalGradePossibleAll: data.original_grade ? data.original_grade.possible_all : null,
          originalGradeEarnedGraded: data.original_grade
            ? data.original_grade.earned_graded : null,
          originalGradePossibleGraded: data.original_grade
            ? data.original_grade.possible_graded : null,
        }));
      } else {
        dispatch(errorFetchingGradeOverrideHistory(data.error_message));
      }
    })
    .catch(() => {
      dispatch(errorFetchingGradeOverrideHistory(GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG));
    })
);

const fetchMatchingUserGrades = (
  courseId,
  searchText,
  cohort,
  track,
  assignmentType,
  showSuccess,
  options = {},
) => {
  const newOptions = { ...options, searchText, showSuccess };
  return fetchGrades(courseId, cohort, track, assignmentType, newOptions);
};

const fetchPrevNextGrades = (endpoint, courseId, cohort, track, assignmentType) => (
  (dispatch) => {
    dispatch(startedFetchingGrades());
    return getAuthenticatedHttpClient().get(endpoint)
      .then(response => response.data)
      .then((data) => {
        dispatch(gotGrades({
          grades: data.results.sort(sortAlphaAsc),
          cohort,
          track,
          assignmentType,
          prev: data.previous,
          next: data.next,
          courseId,
          totalUsersCount: data.total_users_count,
          filteredUsersCount: data.filtered_users_count,
        }));
        dispatch(finishedFetchingGrades());
      })
      .catch(() => {
        dispatch(errorFetchingGrades());
      });
  }
);

const updateGrades = (courseId, updateData, searchText, cohort, track) => (
  (dispatch) => {
    dispatch(gradeUpdateRequest());
    return LmsApiService.updateGradebookData(courseId, updateData)
      .then(response => response.data)
      .then((data) => {
        dispatch(gradeUpdateSuccess(courseId, data));
        dispatch(fetchMatchingUserGrades(
          courseId,
          searchText,
          cohort,
          track,
          defaultAssignmentFilter,
          true,
          { searchText },
        ));
      })
      .catch((error) => {
        dispatch(gradeUpdateFailure(courseId, error));
      });
  }
);

const submitFileUploadFormData = (courseId, formData) => (
  (dispatch) => {
    dispatch(startedCsvUpload());
    return LmsApiService.uploadGradeCsv(courseId, formData).then(() => {
      dispatch(finishedCsvUpload());
      dispatch(uploadOverrideSuccess(courseId));
    }).catch((err) => {
      dispatch(uploadOverrideFailure(courseId, err));
      if (err.status === 200 && err.data.error_messages.length) {
        const { error_messages: errorMessages, saved, total } = err.data;
        return dispatch(csvUploadError({ errorMessages, saved, total }));
      }
      return dispatch(csvUploadError({ errorMessages: ['Unknown error.'] }));
    });
  }
);

const fetchBulkUpgradeHistory = courseId => (
  // todo add loading effect
  dispatch => LmsApiService.fetchGradeBulkOperationHistory(courseId).then(
    (response) => { dispatch(gotBulkHistory(response)); },
  ).catch(() => dispatch(bulkHistoryError()))
);

const updateGradesIfAssignmentGradeFiltersSet = (
  courseId,
  cohort,
  track,
  assignmentType,
) => (dispatch, getState) => {
  const { filters } = getState();
  const hasAssignmentGradeFiltersSet = filters.assignmentGradeMax || filters.assignmentGradeMin;
  if (hasAssignmentGradeFiltersSet) {
    dispatch(fetchGrades(
      courseId,
      cohort,
      track,
      assignmentType,
    ));
  }
};

export {
  startedFetchingGrades,
  finishedFetchingGrades,
  errorFetchingGrades,
  gotGrades,
  fetchGrades,
  fetchMatchingUserGrades,
  fetchPrevNextGrades,
  gradeUpdateRequest,
  gradeUpdateSuccess,
  gradeUpdateFailure,
  updateGrades,
  toggleGradeFormat,
  filterAssignmentType,
  closeBanner,
  submitFileUploadFormData,
  fetchBulkUpgradeHistory,
  doneViewingAssignment,
  fetchGradeOverrideHistory,
  updateGradesIfAssignmentGradeFiltersSet,
  downloadBulkGradesReport,
  downloadInterventionReport,
};
