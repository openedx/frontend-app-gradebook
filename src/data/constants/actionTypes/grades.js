const STARTED_FETCHING_GRADES = 'STARTED_FETCHING_GRADES';
const FINISHED_FETCHING_GRADES = 'FINISHED_FETCHING_GRADES';
const ERROR_FETCHING_GRADES = 'ERROR_FETCHING_GRADES';
const GOT_GRADES = 'GOT_GRADES';
const DONE_VIEWING_ASSIGNMENT = 'DONE_VIEWING_ASSIGNMENT';
const GOT_GRADE_OVERRIDE_HISTORY = 'GOT_GRADE_OVERRIDE_HISTORY';
const ERROR_FETCHING_GRADE_OVERRIDE_HISTORY = 'ERROR_FETCHING_GRADE_OVERRIDE_HISTORY';

const FILTER_SELECTED = 'FILTER_SELECTED';
const GRADE_OVERRIDE = 'GRADE_OVERRIDE';
const REPORT_DOWNLOADED = 'REPORT_DOWNLOADED';
const UPLOAD_OVERRIDE = 'UPLOAD_OVERRIDE';
const UPLOAD_OVERRIDE_ERROR = 'UPLOAD_OVERRIDE_ERROR';

const GRADE_UPDATE_REQUEST = 'GRADE_UPDATE_REQUEST';
const GRADE_UPDATE_SUCCESS = 'GRADE_UPDATE_SUCCESS';
const GRADE_UPDATE_FAILURE = 'GRADE_UPDATE_FAILURE';

const TOGGLE_GRADE_FORMAT = 'TOGGLE_GRADE_FORMAT';
const FILTER_BY_ASSIGNMENT_TYPE = 'FILTER_BY_ASSIGNMENT_TYPE';
const CLOSE_BANNER = 'CLOSE_BANNER';
const OPEN_BANNER = 'OPEN_BANNER';

const START_UPLOAD = 'START_UPLOAD';
const UPLOAD_COMPLETE = 'UPLOAD_COMPLETE';
const UPLOAD_ERR = 'UPLOAD_ERR';
const GOT_BULK_HISTORY = 'GOT_BULK_HISTORY';
const BULK_HISTORY_ERR = 'BULK_HISTORY_ERR';
const BULK_GRADE_REPORT_DOWNLOADED = 'BULK_GRADE_REPORT_DOWNLOADED';
const INTERVENTION_REPORT_DOWNLOADED = 'INTERVENTION_REPORT_DOWNLOADED';

export {
  STARTED_FETCHING_GRADES,
  FINISHED_FETCHING_GRADES,
  ERROR_FETCHING_GRADES,
  GOT_GRADES,
  GRADE_UPDATE_REQUEST,
  GRADE_UPDATE_SUCCESS,
  GRADE_UPDATE_FAILURE,
  TOGGLE_GRADE_FORMAT,
  FILTER_BY_ASSIGNMENT_TYPE,
  OPEN_BANNER,
  CLOSE_BANNER,
  START_UPLOAD,
  UPLOAD_COMPLETE,
  UPLOAD_ERR,
  GOT_BULK_HISTORY,
  BULK_HISTORY_ERR,
  DONE_VIEWING_ASSIGNMENT,
  GOT_GRADE_OVERRIDE_HISTORY,
  ERROR_FETCHING_GRADE_OVERRIDE_HISTORY,
  FILTER_SELECTED,
  GRADE_OVERRIDE,
  REPORT_DOWNLOADED,
  UPLOAD_OVERRIDE,
  UPLOAD_OVERRIDE_ERROR,
  BULK_GRADE_REPORT_DOWNLOADED,
  INTERVENTION_REPORT_DOWNLOADED,
};
