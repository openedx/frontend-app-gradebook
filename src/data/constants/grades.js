import { StrictDict } from 'utils';

const EMAIL_HEADING = 'Email';
const TOTAL_COURSE_GRADE_HEADING = 'Total Grade (%)';
const USERNAME_HEADING = 'Username';
const FULL_NAME_HEADING = 'Full Name';

const GradeFormats = StrictDict({
  absolute: 'absolute',
  percent: 'percent',
});

const Headings = StrictDict({
  email: EMAIL_HEADING,
  totalGrade: TOTAL_COURSE_GRADE_HEADING,
  username: USERNAME_HEADING,
  fullName: FULL_NAME_HEADING,
});

export {
  EMAIL_HEADING,
  TOTAL_COURSE_GRADE_HEADING,
  USERNAME_HEADING,
  FULL_NAME_HEADING,
  GradeFormats,
  Headings,
};
