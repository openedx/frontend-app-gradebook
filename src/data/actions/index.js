import { StrictDict } from '../../utils';

import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import config from './config';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  config,
  filters,
  grades,
  roles,
  tracks,
});
