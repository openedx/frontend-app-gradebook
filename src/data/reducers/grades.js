import actions from '../actions/grades';
import filterActions from '../actions/filters';

const initialState = {
  results: [],
  gradeOverrideHistoryResults: [],
  gradeOverrideCurrentEarnedAllOverride: null,
  gradeOverrideCurrentPossibleAllOverride: null,
  gradeOverrideCurrentEarnedGradedOverride: null,
  gradeOverrideCurrentPossibleGradedOverride: null,
  gradeOriginalEarnedAll: null,
  gradeOriginalPossibleAll: null,
  gradeOriginalEarnedGraded: null,
  gradeOriginalPossibleGraded: null,
  headings: [],
  startedFetching: false,
  finishedFetching: false,
  errorFetching: false,
  overrideHistoryError: '',
  gradeFormat: 'percent',
  showSuccess: false,
  prevPage: null,
  nextPage: null,
  showSpinner: true,
  bulkManagement: {},
  totalUsersCount: 0,
  filteredUsersCount: 0,
  isImportGradesActive: false,
  isDownloadInterventionsActive: false,
  isDownloadGradesActive: false,
};

const grades = (state = initialState, { type, payload }) => {
  switch (type) {
    case actions.banner.open.toString():
      return {
        ...state,
        showSuccess: true,
      };
    case actions.banner.close.toString():
      return {
        ...state,
        showSuccess: false,
      };
    case actions.bulkHistory.received.toString():
      return {
        ...state,
        bulkManagement: {
          ...state.bulkManagement,
          history: payload,
        },
      };
    case actions.csvUpload.started.toString(): {
      const { errorMessages, uploadSuccess, ...rest } = state.bulkManagement;
      return {
        ...state,
        showSpinner: true,
        bulkManagement: rest,
      };
    }
    case actions.csvUpload.finished.toString():
      return {
        ...state,
        showSpinner: false,
        bulkManagement: {
          ...state.bulkManagement,
          uploadSuccess: true,
        },
      };
    case actions.csvUpload.error.toString():
      return {
        ...state,
        showSpinner: false,
        bulkManagement: {
          ...state.bulkManagement,
          ...payload,
        },
      };
    case actions.doneViewingAssignment.toString(): {
      const {
        gradeOverrideHistoryResults,
        gradeOverrideCurrentEarnedAllOverride,
        gradeOverrideCurrentPossibleAllOverride,
        gradeOverrideCurrentEarnedGradedOverride,
        gradeOverrideCurrentPossibleGradedOverride,
        gradeOriginalEarnedAll,
        gradeOriginalPossibleAll,
        gradeOriginalEarnedGraded,
        gradeOriginalPossibleGraded,
        ...rest
      } = state;
      return rest;
    }
    case actions.fetching.started.toString():
      return {
        ...state,
        startedFetching: true,
        finishedFetching: false,
        showSpinner: true,
      };
    case actions.fetching.error.toString():
      return {
        ...state,
        finishedFetching: true,
        errorFetching: true,
      };
    case actions.fetching.received.toString():
      return {
        ...state,
        results: payload.grades,
        headings: payload.headings,
        finishedFetching: true,
        errorFetching: false,
        prevPage: payload.prev,
        nextPage: payload.next,
        showSpinner: false,
        courseId: payload.courseId,
        totalUsersCount: payload.totalUsersCount,
        filteredUsersCount: payload.filteredUsersCount,
      };
    case actions.overrideHistory.received.toString():
      return {
        ...state,
        gradeOverrideHistoryResults: payload.overrideHistory,
        gradeOverrideCurrentEarnedAllOverride: payload.currentEarnedAllOverride,
        gradeOverrideCurrentPossibleAllOverride: payload.currentPossibleAllOverride,
        gradeOverrideCurrentEarnedGradedOverride: payload.currentEarnedGradedOverride,
        gradeOverrideCurrentPossibleGradedOverride: payload.currentPossibleGradedOverride,
        gradeOriginalEarnedAll: payload.originalGradeEarnedAll,
        gradeOriginalPossibleAll: payload.originalGradePossibleAll,
        gradeOriginalEarnedGraded: payload.originalGradeEarnedGraded,
        gradeOriginalPossibleGraded: payload.originalGradePossibleGraded,
        overrideHistoryError: '',
      };
    case actions.overrideHistory.error.toString():
      return {
        ...state,
        finishedFetchingOverrideHistory: true,
        overrideHistoryError: payload,
      };
    case actions.toggleGradeFormat.toString():
      return {
        ...state,
        gradeFormat: payload.target.value,
      };
    case actions.update.request.toString():
      return { ...state, showSpinner: true };
    case actions.update.success.toString():
    case actions.update.failure.toString():
      return { ...state, showSpinner: false };

    case filterActions.update.assignmentType.toString():
      return {
        ...state,
        selectedAssignmentType: payload.filterType,
        headings: payload.headings,
      };
    default:
      return state;
  }
};

export { initialState as initialGradesState };
export default grades;
