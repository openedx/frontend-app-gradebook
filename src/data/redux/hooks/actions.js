import { StrictDict } from 'utils';
import actions from 'data/actions';
import { actionHook } from './utils';

const app = StrictDict({
  useSetLocalFilter: actionHook(actions.app.setLocalFilter),
  useSetSearchValue: actionHook(actions.app.setSearchValue),
  useSetShowImportSuccessToast: actionHook(actions.app.setShowImportSuccessToast),
  useSetView: actionHook(actions.app.setView),
  useCloseModal: actionHook(actions.app.closeModal),
  useSetModalState: actionHook(actions.app.setModalState),
});

const filters = StrictDict({
  useUpdateAssignment: actionHook(actions.filters.update.assignment),
  useUpdateAssignmentLimits: actionHook(actions.filters.update.assignmentLimits),
  useUpdateAssignmentType: actionHook(actions.filters.update.assignmentType),
  useUpdateCohort: actionHook(actions.filters.update.cohort),
  useUpdateCourseGradeLimits: actionHook(actions.filters.update.courseGradeLimits),
  useUpdateIncludeCourseRoleMembers: actionHook(actions.filters.update.includeCourseRoleMembers),
  useUpdateTrack: actionHook(actions.filters.update.track),
  useResetFilters: actionHook(actions.filters.reset),
});

const grades = StrictDict({
  useDoneViewingAssignment: actionHook(actions.grades.doneViewingAssignment),
  useDownloadBulkGradesReport: actionHook(actions.grades.downloadReport.bulkGrades),
  useDownloadInterventionReport: actionHook(actions.grades.downloadReport.intervention),
  useToggleGradeFormat: actionHook(actions.grades.toggleGradeFormat),
  useCloseBanner: actionHook(actions.grades.banner.close),
});

export default StrictDict({
  app,
  filters,
  grades,
});
