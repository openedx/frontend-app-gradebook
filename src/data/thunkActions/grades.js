/* eslint-disable import/no-self-import, import/no-named-as-default-member */
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

import { StrictDict } from 'utils';

import GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG from 'data/constants/errors';

import grades from 'data/actions/grades';
import { sortAlphaAsc } from 'data/actions/utils';
import selectors from 'data/selectors';
import lms from 'data/services/lms';

import * as module from './grades';

const { formatGradeOverrideForDisplay } = selectors.grades;

export const defaultAssignmentFilter = 'All';

export const fetchBulkUpgradeHistory = () => (dispatch) => (
  // todo add loading effect
  lms.api.fetch.gradeBulkOperationHistory().then(
    (response) => { dispatch(grades.bulkHistory.received(response)); },
  ).catch(() => dispatch(grades.bulkHistory.error()))
);

export const fetchGrades = (overrides = {}) => (
  (dispatch, getState) => {
    dispatch(grades.fetching.started());
    const { assignmentType, options } = overrides;
    const cohort = selectors.filters.cohort(getState());
    const track = selectors.filters.track(getState());
    const courseId = selectors.app.courseId(getState());
    const fetchOptions = {
      ...selectors.root.localFilters(getState()),
      ...options,
    };
    return lms.api.fetch.gradebookData(
      fetchOptions.searchText || null,
      cohort,
      track,
      fetchOptions,
    ).then(response => response.data)
      .then((data) => {
        dispatch(grades.fetching.received({
          assignmentType: (assignmentType || selectors.filters.assignmentType(getState())),
          cohort,
          courseId,
          track,
          grades: data.results.sort(sortAlphaAsc),
          prev: data.previous,
          next: data.next,
          totalUsersCount: data.total_users_count,
          filteredUsersCount: data.filtered_users_count,
        }));
        if (fetchOptions.showSuccess) {
          dispatch(grades.banner.open());
        }
        dispatch(grades.fetching.finished());
      })
      .catch(() => {
        dispatch(grades.fetching.error());
      });
  }
);

export const fetchGradesIfAssignmentGradeFiltersSet = () => (
  (dispatch, getState) => {
    if (selectors.filters.areAssignmentGradeFiltersSet(getState())) {
      dispatch(module.fetchGrades());
    }
  }
);

export const fetchGradeOverrideHistory = (subsectionId, userId) => (
  dispatch => lms.api.fetch.gradeOverrideHistory(subsectionId, userId)
    .then(response => response.data)
    .then((data) => {
      if (data.success) {
        dispatch(grades.overrideHistory.received({
          overrideHistory: formatGradeOverrideForDisplay(data.history),
          currentEarnedAllOverride: data.override ? data.override.earned_all_override : null,
          currentPossibleAllOverride: data.override ? data.override.possible_all_override : null,
          currentEarnedGradedOverride: data.override ? data.override.earned_graded_override : null,
          currentPossibleGradedOverride: data.override
            ? data.override.possible_graded_override : null,
          originalGradeEarnedAll: data.original_grade ? data.original_grade.earned_all : null,
          originalGradePossibleAll: data.original_grade ? data.original_grade.possible_all : null,
          originalGradeEarnedGraded: data.original_grade
            ? data.original_grade.earned_graded : null,
          originalGradePossibleGraded: data.original_grade
            ? data.original_grade.possible_graded : null,
        }));
      } else {
        dispatch(grades.overrideHistory.error(data.error_message));
      }
    })
    .catch(() => {
      dispatch(grades.overrideHistory.error(GRADE_OVERRIDE_HISTORY_ERROR_DEFAULT_MSG));
    })
);

export const fetchPrevNextGrades = (endpoint) => (
  (dispatch, getState) => {
    dispatch(grades.fetching.started());
    return getAuthenticatedHttpClient().get(endpoint)
      .then(({ data }) => data)
      .then((data) => {
        dispatch(grades.fetching.received({
          courseId: selectors.app.courseId(getState()),
          cohort: selectors.filters.cohort(getState()),
          track: selectors.filters.track(getState()),
          assignmentType: selectors.filters.assignmentType(getState()),
          grades: data.results.sort(sortAlphaAsc),
          prev: data.previous,
          next: data.next,
          filteredUsersCount: data.filtered_users_count,
          totalUsersCount: data.total_users_count,
        }));
        dispatch(grades.fetching.finished());
      })
      .catch(() => {
        dispatch(grades.fetching.error());
      });
  }
);

export const submitImportGradesButtonData = (formData) => (
  (dispatch, getState) => {
    const courseId = selectors.app.courseId(getState());
    dispatch(grades.csvUpload.started());
    return lms.api.uploadGradeCsv(formData).then(() => {
      dispatch(grades.csvUpload.finished());
      dispatch(grades.uploadOverride.success(courseId));
    }).catch((error) => {
      dispatch(grades.uploadOverride.failure({ courseId, error }));
      if (error.status === 200 && error.data.error_messages.length) {
        const { error_messages: errorMessages, saved, total } = error.data;
        return dispatch(grades.csvUpload.error({ errorMessages, saved, total }));
      }
      return dispatch(grades.csvUpload.error({ errorMessages: ['Unknown error.'] }));
    });
  }
);

export const updateGrades = () => (
  (dispatch, getState) => {
    const updateData = selectors.app.editUpdateData(getState());
    dispatch(grades.update.request());
    return lms.api.updateGradebookData(updateData)
      .then(response => response.data)
      .then((data) => {
        dispatch(grades.update.success({ data }));
        dispatch(module.fetchGrades({
          assignmentType: defaultAssignmentFilter,
          options: { showSuccess: true },
        }));
      })
      .catch((error) => {
        dispatch(grades.update.failure({ error }));
      });
  }
);

export default StrictDict({
  fetchBulkUpgradeHistory,
  fetchGrades,
  fetchGradesIfAssignmentGradeFiltersSet,
  fetchGradeOverrideHistory,
  fetchPrevNextGrades,
  submitImportGradesButtonData,
  updateGrades,
});
