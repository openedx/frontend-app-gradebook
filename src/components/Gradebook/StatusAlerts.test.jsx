import React from 'react';
import { shallow } from 'enzyme';
import {
  StatusAlerts,
  mapDispatchToProps,
  mapStateToProps,
  maxCourseGradeInvalidMessage,
  minCourseGradeInvalidMessage,
} from './StatusAlerts';
import { closeBanner } from '../../data/actions/grades';

jest.mock('@edx/paragon', () => ({
  StatusAlert: 'StatusAlert',
}));

describe('StatusAlerts', () => {
  let props = {
    showSuccessBanner: true,
    isMaxCourseGradeFilterValid: true,
    isMinCourseGradeFilterValid: true,
  };

  beforeEach(() => {
    props = {
      ...props,
      handleCloseSuccessBanner: jest.fn().mockName('handleCloseSuccessBanner'),
    };
  });

  describe('snapshots', () => {
    let el;
    it('basic snapshot', () => {
      el = shallow(<StatusAlerts {...props} />);
      const courseGradeFilterAlertDialogText = 'the quiCk brown does somEthing or other';
      jest.spyOn(
        el.instance(),
        'courseGradeFilterAlertDialogText',
        'get',
      ).mockReturnValue(courseGradeFilterAlertDialogText);
      expect(el.instance().render()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it.each([
      [false, false],
      [false, true],
      [true, false],
      [true, true],
    ])('min + max course grade validity', (isMinCourseGradeFilterValid, isMaxCourseGradeFilterValid) => {
      props = {
        ...props,
        isMinCourseGradeFilterValid,
        isMaxCourseGradeFilterValid,
      };
      const el = shallow(<StatusAlerts {...props} />);
      expect(
        el.instance().isCourseGradeFilterAlertOpen,
      ).toEqual(
        !isMinCourseGradeFilterValid || !isMaxCourseGradeFilterValid,
      );
      if (!isMaxCourseGradeFilterValid) {
        expect(
          el.instance().courseGradeFilterAlertDialogText,
        ).toEqual(
          expect.stringContaining(maxCourseGradeInvalidMessage),
        );
      }
      if (!isMinCourseGradeFilterValid) {
        expect(
          el.instance().courseGradeFilterAlertDialogText,
        ).toEqual(
          expect.stringContaining(minCourseGradeInvalidMessage),
        );
      }
    });
  });

  describe('mapStateToProps', () => {
    it('showSuccessBanner', () => {
      const arbitraryValue = 'AppleBananaCucumber';
      const state = {
        grades: {
          showSuccess: arbitraryValue,
        },
      };
      expect(mapStateToProps(state).showSuccessBanner).toBe(arbitraryValue);
    });
  });
  describe('handleCloseSuccessBanner', () => {
    test('handleCloseSuccessBanner', () => {
      expect(
        mapDispatchToProps.handleCloseSuccessBanner,
      ).toEqual(
        closeBanner,
      );
    });
  });
});
