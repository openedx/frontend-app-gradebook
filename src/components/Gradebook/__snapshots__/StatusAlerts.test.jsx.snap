// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusAlerts snapshots basic snapshot 1`] = `
<React.Fragment>
  <StatusAlert
    alertType="success"
    dialog="The grade has been successfully edited. You may see a slight delay before updates appear in the Gradebook."
    onClose={[MockFunction handleCloseSuccessBanner]}
    open={true}
  />
  <StatusAlert
    alertType="danger"
    dialog="the quiCk brown does somEthing or other"
    dismissible={false}
    open={false}
  />
</React.Fragment>
`;
